packages/dds/tree/src/simple-tree/toMapTree.ts

```diff
@@ -236,94 +181,121 @@ function mapValueWithFallbacks(
 	}
 }
 
-function arrayToMapTree(
-	data: InsertableContent[],
-	globalSchema: FlexTreeSchema,
-	typeSet: AllowedTypeSet,
-): MapTree {
-	const schema = getType(data, globalSchema, typeSet);
-	assert(
-		schema instanceof FlexFieldNodeSchema,
-		0x84b /* Array data reported comparable with the schema without a primary field. */,
-	);
+function arrayToMapTreeFields(data: InsertableContent[], allowedTypes: AllowedTypes): MapTree[] {
+	const mappedData: MapTree[] = [];
+	for (const child of data) {
+		// We do not support undefined sequence entries.
+		// If we encounter an undefined entry, use null instead if supported by the schema, otherwise throw.
+		let childWithFallback = child;
+		if (child === undefined) {
+			if (allowedTypes.includes(nullSchema)) {
+				childWithFallback = null;
+			} else {
+				throw new TypeError(`Received unsupported array entry value: ${child}.`);
+			}
+		}
+		const mappedChild = nodeDataToMapTree(childWithFallback, allowedTypes);
+		mappedData.push(mappedChild);
+	}
 
-	const mappedChildren = fieldDataToMapTrees(data, globalSchema, schema.info);
-	const fieldsEntries: [FieldKey, MapTree[]][] =
-		mappedChildren.length === 0 ? [] : [[EmptyKey, mappedChildren]];
+	return mappedData;
+}
+
+function arrayToMapTree(data: InsertableContent[], allowedTypes: AllowedTypes): MapTree {
+	const schema = getType(data, allowedTypes);
+	if (schema.kind !== NodeKind.Array) {
+		fail(`Provided array input is incompatible with schema "${schema.identifier}".`);
```

Comment:
If we expect this to ever be a user facing error, we should throw a TypeError or UssageError here instead of a fail.